src/diffusers/models/transformers/transformer_2d.py

```diff
@@ -153,94 +149,178 @@ def __init__(
                 f" {patch_size}. Make sure that `in_channels`, `num_vector_embeds` or `num_patches` is not None."
             )
 
-        # 2. Define input layers
-        if self.is_input_continuous:
-            self.in_channels = in_channels
+        self.in_channels = in_channels
 
-            self.norm = torch.nn.GroupNorm(num_groups=norm_num_groups, num_channels=in_channels, eps=1e-6, affine=True)
-            if use_linear_projection:
-                self.proj_in = linear_cls(in_channels, inner_dim)
-            else:
-                self.proj_in = conv_cls(in_channels, inner_dim, kernel_size=1, stride=1, padding=0)
-        elif self.is_input_vectorized:
-            assert sample_size is not None, "Transformer2DModel over discrete input must provide sample_size"
-            assert num_vector_embeds is not None, "Transformer2DModel over discrete input must provide num_embed"
-
-            self.height = sample_size
-            self.width = sample_size
-            self.num_vector_embeds = num_vector_embeds
-            self.num_latent_pixels = self.height * self.width
-
-            self.latent_image_embedding = ImagePositionalEmbeddings(
-                num_embed=num_vector_embeds, embed_dim=inner_dim, height=self.height, width=self.width
+        # 2. Initialize the right blocks.
+        if self.is_input_continuous:
+            self._init_continuous_input(
+                in_channels=in_channels, use_linear_projection=use_linear_projection, norm_type=norm_type
             )
+        elif self.is_input_vectorized:
+            self._init_vectorized_inputs(sample_size=sample_size, norm_type=norm_type)
         elif self.is_input_patches:
-            assert sample_size is not None, "Transformer2DModel over patched input must provide sample_size"
-
-            self.height = sample_size
-            self.width = sample_size
-
-            self.patch_size = patch_size
-            interpolation_scale = (
-                interpolation_scale if interpolation_scale is not None else max(self.config.sample_size // 64, 1)
-            )
-            self.pos_embed = PatchEmbed(
-                height=sample_size,
-                width=sample_size,
-                patch_size=patch_size,
+            self._init_patched_inputs(
                 in_channels=in_channels,
-                embed_dim=inner_dim,
+                sample_size=sample_size,
+                norm_type=norm_type,
+                patch_size=patch_size,
                 interpolation_scale=interpolation_scale,
+                caption_channels=caption_channels,
             )
 
-        # 3. Define transformers blocks
+        # Others.
+        self.out_channels = in_channels if out_channels is None else out_channels
+        self.gradient_checkpointing = False
+
+    def _init_continuous_input(self, in_channels, use_linear_projection, norm_type):
```

Comment:
do we need to pass any of the arguments? they are all accessible via `self.config`, no?